Safewaan/HonkedIn
# Users can join events, Event creators can set max participants

## Trello ticket link

[As a user, I want to be able to join events other users have created.](https://trello.com/c/U513rOy9/107-as-a-user-i-want-to-be-able-to-join-events-other-users-have-created)

[As a user, I want to be able to limit the number of users who can attend and RSVP to my events.](https://trello.com/c/WntwvEI7/53-as-a-user-i-want-to-be-able-to-limit-the-number-of-users-who-can-attend-and-rsvp-to-my-events)



## Implementation description
1. General Changes
* Added packages for `@mui/material`
* Updated existing packages to support `@mui/material`
* Removed some unused code leftover in `CreateEvent.js` and `Events.js`
* Removed comments @gabbytang, they looked like possible implementations of code that weren't used. Let me know if they should be restored. 
* Updated the way that the `userID` is obtained through email from Firebase. I noticed some issues with react states being delayed (for example, updating a state then using the state immediately will not reflect the changes made). This caused some issues with requests being sent with null a `userID`.
* Fixed spacing

2. `CreateEvent.js` and `events` table Changes
* Added more columns in the `events` table -> participants, totalParticipants and status
* Added field for participants in "Create Event" page
* Updated `createEvent` API to take the `totalParticipants` value for the event
* Removed the date check from `getEvents` as events that have passed should still be displayed, but their status should be updated respectively (this will need to be implemented in a future PR)

3. `Events.js` Changes
* Created a new table, `eventParticipants` that stores the the participantIDs of the event. 
* Created a new API, `joinEvents`, that allows a user to join a given event, only if the event is not full and the user has not already joined the event.
* Updated the card component for events to include a button where the user can view more details about the event. 
* Added a popup that appears when a user clicks "View Event" that provides more info on the event, and allows users to join the event.
* Added a toast message that informs the user if they've successfully joined the event, have already joined the event, or if the event is full.
* Increased the width of the card component. Note: it is not centred but can be changed later. 



## Steps to test
0. Setup
*  install the new dependencies with `npm i`

1. `CreateEvent.js` Tests
* Verify that the "Event Participants" field throws an error if the form is submitted without the field being filled in.
* Verify that if an event is created and all fields are inputted correctly, the corresponding row appears in the `events` table with the matching `totalParticipants` value from the form.
* Verify that the `participants` value is set to 0 and that the `status` is set to "Active".

2. `Events.js` Tests
* Open up the `events` page and verify the events load correctly.
* Verify that the "View Event" button can be clicked, and the corresponding event's details are displayed in a popup. 
* Verify that the popup can be closed with the "Close" button. 
* Verify that an event can be joined **given that the event has not already been joined**. Verify that a success toast message appears and that the page is refreshed after a short delay. Verify that a corresponding row appears in `eventParticipants`.
* Verify that an event that has already been joined by the user cannot be joined. Verify that an error toast message appears. Verify that a new row **does not** appear in `eventParticipants`.
* Verify that an event at max capacity cannot be joined. Please use the test event, "Max Participant Test" to test this. Verify that an error toast message appears. Verify that a new row **does not** appear in `eventParticipants`. 



## What should reviewers focus on?
* I've added catches for error cases (in the form of toast messages) but if you can think of any ways to break the code that I haven't included in the test steps, please experiment and let me know if anything breaks.
